network_analysis/nbconverted/2B_community_detection_louvain.py

```diff
@@ -0,0 +1,173 @@
+
+# map of Pa gene names to generic/not generic status, generated by Alex
+generic_gene_map = os.path.join('..', 'pseudomonas_analysis', 'annot_df.tsv')
+
+# script parameters
+NUM_NODE_SAMPLES = 1000 # number of degree-matched node samples for permutation test
+NUM_BINS = 100 # number of bins to divide nodes into, for sampling
+
+
+# In[3]:
+
+
+graph_df = pd.read_csv(edge_list)
+graph_df.head()
+
+
+# In[4]:
+
+
+G = ig.Graph.TupleList(graph_df.values,
+                       weights=True,
+                       directed=False)
+
+
+# In[5]:
+
+
+# make sure vertex/edge properties exist
+print(G.es['weight'][:5])
+
+
+# In[6]:
+
+
+annot_df = pd.read_csv(generic_gene_map, sep='\t', index_col=0)
+annot_df.head()
+
+
+# In[7]:
+
+
+is_generic = [int(annot_df.loc[v['name'], 'label']) for v in G.vs]
+G.vs['is_generic'] = is_generic
+
+
+# In[8]:
+
+
+# community detection using Louvain modularity optimization
+partition = G.community_multilevel(weights=G.es['weight'], return_levels=False)
+# plot?
+
+
+# In[9]:
+
+
+# get dataframe mapping Pa genes to communities
+def graph_partition_to_df(G, partition):
+    clusters = []
+    for label, vl in enumerate(partition):
+        clusters += [(G.vs['name'][v],
+                      label,
+                      G.degree(v),
+                      G.vs['is_generic'][v]) for v in vl]
+    return pd.DataFrame(clusters,
+                        columns=['gene', 'label', 'degree', 'is_generic'])
+
+labels_df = graph_partition_to_df(G, partition)
+print(len(labels_df.label.unique()))
+labels_df.sort_values(by='degree', ascending=False).head()
+
+
+# In[10]:
+
+
+# simultaneously sort nodes and degrees by degree, ascending
+nodes, degrees, is_generic = sort_by_degree(labels_df.gene.values,
+                                            labels_df.degree.values,
+                                            labels_df.is_generic.values)
+
+# sample a few times and add results to df
+for it in range(NUM_NODE_SAMPLES):
+    s_nodes, s_degrees, __ = sample_degree_matched(nodes, degrees, is_generic,
+                                                   num_bins=NUM_BINS)
+    sampled = [1 if gene in s_nodes else 0 for gene in labels_df.gene]
+    labels_df['sampled_{}'.format(it)] = sampled
+
+labels_df.sort_values(by='degree', ascending=False).iloc[:5, :5]
+
+
+# In[11]:
+
+
+generic_count_df = (
+    labels_df.groupby('label').sum()
+             .drop(columns=['degree'])
+)
+print(generic_count_df.shape)
+generic_count_df.sort_values(by='is_generic', ascending=False).iloc[:5, :5]
+
+
+# In[12]:
+
+
+nonzero_counts_df = pd.DataFrame(
+    [np.count_nonzero(generic_count_df, axis=0)],
+    columns=generic_count_df.columns
+)
+nonzero_counts_df.iloc[:, :5]
+
+
+# In[13]:
+
+
+n_generic_groups = nonzero_counts_df.iloc[0, 0]
+n_total_groups = len(labels_df.label.unique())
+
+sns.set({'figure.figsize': (8, 5)})
+
+sns.histplot(nonzero_counts_df.iloc[0, 1:].values,
```

Comment:
Just curious, is there a reason why the Louvain produces more modules?

Reply:
My guess would just be overfitting; i.e. many of the small communities should really be a single larger community. [Figure 1 of this paper](https://journals.aps.org/prresearch/pdf/10.1103/PhysRevResearch.2.043271) shows that methods like Louvain (the "Q=" methods) tend to find more communities than the true number of communities in generated graphs.

People have a lot of strong opinions about this sort of thing, though. I picked Louvain because empirically it seems to do pretty well in biological networks: e.g. in [our DREAM challenge paper](https://doi.org/10.1038/s41592-019-0509-5) many of the top performing methods used some variation of Louvain. It's also fairly straightforward to understand. But I also think the more statistically principled models, like the planted partition model described in the first paper I linked above, have some advantages.

In any case, the results here don't seem to be sensitive to the choice of algorithm, which makes me more confident that what we're seeing is a real signal.